internal/auth/internal/mappers/trafficpermissionsmapper/traffic_permissions_mapper.go

```diff
@@ -0,0 +1,162 @@
+	lock                     sync.Mutex
+	mapper                   *bimapper.Mapper
+	missingMap               *MissingWorkloadIdentityMapper // indexes on the name of the missing Workload Identity
+	workloadIdentityToCTPMap map[string]*pbresource.ID
+}
+
+// TODO: We wouldn't need this if we can make a generic bimapper that doesn't clean up dead-end links.
+// MissingWorkloadIdentityMapper holds traffic permissions with explicit destinations that are missing an active workload identity.
+type MissingWorkloadIdentityMapper struct {
+	tpToWorkloadIdentityName map[resource.ReferenceKey]map[string]bool
+	workloadIdentityNameToTP map[string]map[resource.ReferenceKey]bool
+}
+
+func newMissingWorkloadIdentityMapper() *MissingWorkloadIdentityMapper {
+	return &MissingWorkloadIdentityMapper{
+		tpToWorkloadIdentityName: make(map[resource.ReferenceKey]map[string]bool),
+		workloadIdentityNameToTP: make(map[string]map[resource.ReferenceKey]bool),
+	}
+}
+
+func (mm *MissingWorkloadIdentityMapper) track(tp *pbresource.ID, name string) {
+	tpRef := resource.NewReferenceKey(tp)
+	_, ok := mm.workloadIdentityNameToTP[name]
+	if !ok {
+		mm.workloadIdentityNameToTP[name] = make(map[resource.ReferenceKey]bool)
+	}
+	mm.workloadIdentityNameToTP[name][tpRef] = true
+	_, ok = mm.tpToWorkloadIdentityName[tpRef]
+	if !ok {
+		mm.tpToWorkloadIdentityName[tpRef] = make(map[string]bool)
+	}
+	mm.tpToWorkloadIdentityName[tpRef][name] = true
+}
+
+func (mm *MissingWorkloadIdentityMapper) untrack(tp *pbresource.ID) {
+	tpRef := resource.NewReferenceKey(tp)
+	wiNames, ok := mm.tpToWorkloadIdentityName[tpRef]
+	if ok {
+		for wiName := range wiNames {
+			delete(mm.workloadIdentityNameToTP[wiName], tpRef)
+		}
+	}
+	delete(mm.tpToWorkloadIdentityName, tpRef)
+}
+
+func New() *TrafficPermissionsMapper {
+	return &TrafficPermissionsMapper{
+		lock:                     sync.Mutex{},
+		mapper:                   bimapper.New(types.ComputedTrafficPermissionsType, types.TrafficPermissionsType),
+		missingMap:               newMissingWorkloadIdentityMapper(),
+		workloadIdentityToCTPMap: make(map[string]*pbresource.ID, 0),
+	}
+}
+
+//  TrafficPermissionsMapper functions
+
+func (tm *TrafficPermissionsMapper) MapTrafficPermissions(_ context.Context, _ controller.Runtime, res *pbresource.Resource) ([]controller.Request, error) {
+	tm.lock.Lock()
+	defer tm.lock.Unlock()
+	workloadIdentities := make([]controller.Request, 0)
+	// get already associated WorkloadIdentities/CTPs for reconcile requests:
+	associatedWIs := tm.mapper.ItemIDsForLink(res.Id)
+	for _, w := range associatedWIs {
+		workloadIdentities = append(workloadIdentities, controller.Request{ID: w})
+	}
+	var tp pbauth.TrafficPermissions
+	err := res.Data.UnmarshalTo(&tp)
+	if err != nil {
+		return nil, resource.NewErrDataParse(&tp, err)
+	}
+	newWorkloadIdentities := make([]controller.Request, 0)
+	// get new CTP associations based on destination
+	if len(tp.Destination.IdentityName) > 0 {
```

Comment:
Are we ignoring `IdentityPrefix` for now? It looks like it isn't supported by [validations](https://github.com/hashicorp/consul/blob/5d99fb7bdf158b9b3771c4f9defc0fdf1cd60d8c/internal/auth/internal/types/traffic_permissions.go#L52) yet, but it feels like it could have a big impact on this logic.

Reply:
FYI I asked Sarah to remove identity prefix because it's not something we proposed in the RFC.